### ISSUE TITLE
Typecast error when trying to pass a subtype of a record with a nested record into a function with a typedesc feature

### ISSUE THREAD

**Issue by sachinira** [ballerina-platform/ballerina-lang#32653]
**Description:**
We are trying to implement an external function which infers typedesc value of the `returnType` from LHS as shown below.

```ballerina
import ballerina/jballerina.java;

public type CountryVariables record {|
    string code;
|};

public type Country record {|
    string code?;
    string name?;
    string native?;
    string phone?;
    Continent continent?;
    string? capital?;
    string? currency?;
    string emoji?;
    string emojiU?;

|};

public type Continent record {|
    string code?;
    string name?;
|};

public isolated client class Client {
    remote isolated function country(typedesc<Country> returnType = <>, *CountryVariables variables)
    returns returnType = @java:Method {
        'class: "io.ballerina.graphql.nativeimpl.QueryProcessor",
        name: "nativeQuery"
    } external;
}
```

```java
public static BMap<BString, Object> nativeQuery(Environment env, BObject client, BTypedesc ballerinaType,
                                                    BMap<BString, Object> variables) {
        ...
}
```

```ballerina
Client baseClient = new Client();

type MyCountry record {|
    string code?;
    string phone?;
    record {|
        string code?;
    |} continent?;
|};

function testCountry() {
    MyCountry response = baseClient->country(code = "LK");
    log:printInfo(response.toString());
}
```

This causes a typecast error 
```
{ballerina}TypeCastError {"message":"incompatible types: 'grapql:MyCountry' cannot be cast to 'grapql:Country'"}
```

**Steps to reproduce:**

```ballerina
import ballerina/io;

public type Country record {|
    string code?;
    string name?;
    string native?;
    string phone?;
    Continent continent?;
    string? capital?;
    string? currency?;
    string emoji?;
    string emojiU?;

|};

public type Continent record {|
    string code?;
    string name?;
|};


type MyCountry record {|
    string code?;
    string phone?;
    record {|
        string code?;
    |} continent?;
|};

public function main() {
    MyCountry x = {};
    Country y = x;
    io:println(x is Country);
}
```
Run this code

```
Compiling source
	test.bal
HINT [test.bal:(33:16,33:28)] unnecessary condition: expression will always evaluate to 'true'

Running executable

false
```

**Affected Versions:**
Ballerina SL Beta2

**OS, DB, other environment details and versions:**
JDK 11
Ubuntu 20.04 LTS

**Comment by MaryamZi** [comment]
This is due to a bug in the type-checking logic for records with optional fields.

The check at https://github.com/ballerina-platform/ballerina-lang/blob/master/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java#L1160 should not return false if the field is optional in the target type.

The check should be updated to be similar to the compile-time check https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L2989.

**Comment by Nadeeshan96** [comment]
Fixed with #32729.
